### PR TITLE
Demo 2521 - Adding the new footnote feature to the deployer.

### PR DESCRIPTION
**Pull request by nr-kkenney**
The footnote field is a new field in the deployment config that is a mostly free form text field that will be printed out at the end of a deployment. 

The field can accept either a String, or Array<String> value. 
- For an array<string> value, the text is concatenated with '\n' between indices. This is meant for ease of use and visual organization.
- For a string value, the text within displayed as is (except for the exception mentioned below).

I mentioned above that a 'footnote' is mostly freeform, with the exception being that merged fields can be added within the text of a footnote, and their appropriate values will be substituted before being displayed.

Additionally, there is validation of the footnote on deployment, and if this fails, the deployment fails with an error message. The validation is fairly basic: just ensuring that the footnote is of type String, or Array<String>.

### CODE REVIEW

File: src/footnote/orchestrator.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+      @json = json || Common::JsonParser.new
+      @logger = logger || Common::Logger::LoggerFactory.get_logger
+    end
+
+    # Method that will print out the content of the 'footnote' field in the deployment config, with merge fields substituted.
+    # @return [nil]
+    def execute
+      deploy_config = @context.get_command_line_provider.get_deployment_config_content
+      raw_footnote = @json.get(deploy_config, 'footnote')
+      string_footnote = convert_to_string(raw_footnote)
```
Comment (aswanson-nr) [inline src/footnote/orchestrator.rb:21]: You should be able to run `field_merger.merge_values(string_footnote)`, no need to filter out the fields.

Comment (nr-kkenney) [reply src/footnote/orchestrator.rb:21]: thats a good callout. i remember seeing this, but totally overlooked it.

File: src/orchestrator.rb [hunk 2]
```diff
@@ -10,7 +10,7 @@ def initialize(
       context,
       configuration_orchestrator = nil, provision_orchestrator = nil,
       install_orchestrator = nil, teardown_orchestrator = nil,
-      summary_orchestrator = nil)
+      summary_orchestrator = nil, footnote_orchestrator = nil)
```
Comment (nr-kkenney) [inline src/orchestrator.rb:13]: remove footnote_orchestrator

**Review by aswanson-nr — APPROVED** [review]

File: src/command_line/provider.rb [hunk 3]
```diff
@@ -66,6 +66,14 @@ def get_logging_level()
       return @options[:logging_level].downcase()
     end
 
+    # Method to return deployment config either as an OpenStruct or Hash.
```
Comment (gabeobrien) [inline src/command_line/provider.rb:69]: I am personally not a huge fan of these type of comments.  I do not think they add much.  The name of the method here (get_deployment_config) seem pretty clear.  The comment just add more text to the file with little value added.

Comment (nr-kkenney) [reply src/command_line/provider.rb:69]: just for this particular case, or do you mean all the comments added?

Comment (gabeobrien) [reply src/command_line/provider.rb:69]: In general I am not a fan of comment that just restate what the class/method does.  The class/method should be named in a way that what it is doing is clear.  Comment (in my opinion) should be left for areas of the code that are not clear and could be confusing.  To me most (if not all) of the comments (outside of the ones you added to give hinting to Visual Studio Code) are not needed.

Comment (aswanson-nr) [reply src/command_line/provider.rb:69]: This is part of the information needed for the VS code extension or any other language extension. JSDoc follows the same pattern of description, parameters, and then return value. While I agree that the function name should convey this information, this is a standard way to document functions in an interpreted language like python, ruby, and javascript. I think we should leave this as is, especially if we would like to generate documentation someday.

Comment (aswanson-nr) [reply src/command_line/provider.rb]: Removing comments for now, created future MMF for that work.

**Pull request merged**